Σε αυτό το Notebook παρουσιάζεται η λύση της πρώτης εργασίας για το μάθημα Ανάλυση Δικτύων και Εξόρυξη Γνώσης από τον Παγκόσμιο Ιστό  

In [1]:
#Importing Libraries
import networkx as nx
from networkx.algorithms.reciprocity import reciprocity
import matplotlib.pyplot as plt
import numpy as np
import matplotlib as mpl
import warnings
%matplotlib inline
warnings.filterwarnings('ignore')
# change defaults to be less ugly
mpl.rc('xtick', labelsize=14, color="#222222") 
mpl.rc('ytick', labelsize=14, color="#222222") 
mpl.rc('xtick.major', size=6, width=1)
mpl.rc('xtick.minor', size=3, width=1)
mpl.rc('ytick.major', size=6, width=1)
mpl.rc('ytick.minor', size=3, width=1)
mpl.rc('axes', linewidth=1, edgecolor="#222222", labelcolor="#222222")
mpl.rc('text', usetex=False, color="#222222")
import warnings
warnings.filterwarnings('ignore')

# First Section (Gnutella Network) 



In [ ]:
G1 = nx.read_edgelist("p2p-Gnutella05.txt",create_using=nx.DiGraph(), nodetype = int)

print('Basic Analysis for Gnutella Network (G1)')
print()
#nodes - edges 
print('Number of Nodes for G1 is: ' + str(G1.number_of_nodes())) 
print('Number of Edges for G1 is: ' + str(G1.number_of_edges())) 

#Nodes with Self Loops 
print('Number of Nodes with self loops is: ' + str(len(list(nx.nodes_with_selfloops(G1)))))

# Number of Reciprocated Edges (Το documentation του networkx μας αναφέρει ότι η συνάρτηση reciprocity υπολογίζει τον λόγο των ακμών που δείχνουν και στις δύο 
# κατευθήνσεις, προς το πλήθος όλων των ακμών του δικτύου. Οπότε για να υπολογίζουμε το πλήθος των reciprocated edges αρκεί να πολλαπλασιάσουμε το αποτέλεσμα
# που παράγει η συνάρτηση reciprocity με το πλήθος όλων των ακμών) 
print('Number of reciprocated edges is: ' + str(reciprocity(G1)*G1.number_of_edges()))

# #Number of Sink and Source nodes 
print('Number of Sink Nodes in G1: ' + str(len([node for node in G1.nodes if G1.out_degree(node) == 0])))
print('Number of Source Nodes in G1: ' + str(len([node for node in G1.nodes if G1.in_degree(node) == 0])))


#Max, Min and Average In-going Degree
ingoing_degrees_G1 = dict(G1.in_degree())
avg_ingoing_degree_G1 = sum(val for key, val in ingoing_degrees_G1.items()) / G1.number_of_nodes()
print('Max in-going Degree of a Node: ' + str(max(ingoing_degrees_G1.values()))) 
print('Min in-going Degree of a Node: ' + str(min(ingoing_degrees_G1.values())))
print('Average in-going degree of Nodes: ' + str(avg_ingoing_degree_G1)) 

# #Max, Min and Average Out-going Degree
outgoing_degrees_G1 = dict(G1.out_degree())
avg_outgoing_degree_G1 = sum(val for key, val in outgoing_degrees_G1.items()) / G1.number_of_nodes()
print('Max Out-going Degree of a Node: ' + str(max(outgoing_degrees_G1.values()))) 
print('Min Out-going Degree of a Node: ' + str(min(outgoing_degrees_G1.values())))
print('Average Out-going degree of Nodes: ' + str(avg_outgoing_degree_G1)) 

#Diameter of the Network (δεν είναι συνεκτικά συνδεδεμένο, οπότε θα βρω την μεγαλύτερη διάμετρο από τις συνιστώσες του. Επίσης δεν λαμβάνω υπόψη την φορά των συνιστωσών του
# για να βρεθεί το ίδιο αποτέλεσμα όπως την ιστοσελίδα από την οποία κατέβασα το network)
components = list(nx.weakly_connected_components(G1))
max_diameter = 0
for component in components:
    diameter = nx.diameter(G1.subgraph(component).to_undirected())
    if diameter > max_diameter:
        max_diameter = diameter
print("The diameter is:", max_diameter)

#Average clustering coefficient and global clustering coefficient 
clustering_coefficients = nx.clustering(G1)
avg_clustering_coefficient = sum(val for key,val in clustering_coefficients.items())/G1.number_of_nodes()
print('Average Clustering Coefficient is: ' + str(avg_clustering_coefficient*2)) 
#Για να βγει το σωστό αποτέλεσμα το πολλαπλασιάζω *2 

global_clustering_coefficient = nx.transitivity(G1)
print('Global Clustering Coefficient is ' + str(global_clustering_coefficient*2)) 

#strongly connected component
scc_G1 = list(nx.strongly_connected_components(G1))
largest_scc_G1 = max(scc_G1, key=len)
print('Number of Nodes in the largest Strongly connected Component: ' + str(len(largest_scc_G1))) 
num_edges_largest_scc_G1 = G1.subgraph(largest_scc_G1).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_scc_G1))

#Weakly connected component 
#weekly connected 
wcc_G1 = list(nx.weakly_connected_components(G1))

largest_wcc_G1 = max(wcc_G1, key=len)
print('Number of Nodes in the largest Weakly connected Component: ' + str(len(largest_wcc_G1))) 
num_edges_largest_wcc_G1 = G1.subgraph(largest_wcc_G1).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_wcc_G1))


Basic Analysis for Gnutella Network (G1)

Number of Nodes for G1 is: 8846
Number of Edges for G1 is: 31839
Number of Nodes with self loops is: 0
Number of reciprocated edges is: 0.0
Number of Sink Nodes in G1: 4996
Number of Source Nodes in G1: 118
Max in-going Degree of a Node: 79
Min in-going Degree of a Node: 0
Average in-going degree of Nodes: 3.5992539000678274
Max Out-going Degree of a Node: 65
Min Out-going Degree of a Node: 0
Average Out-going degree of Nodes: 3.5992539000678274


KeyboardInterrupt: ignored

# Second Section (General Relativity and Quantum Cosmology)

In [ ]:
G2 = nx.read_edgelist("CA-GrQc.txt",create_using=nx.Graph(), nodetype = int)

print('Basic Analysis for General Realtivity and Quantum Cosmology Network (G2)')
print()
#nodes - edges 
print('Number of Nodes for G2 is: ' + str(G2.number_of_nodes())) 
print('Number of Edges for G2 is: ' + str(G2.number_of_edges())) 

#Nodes with Self Loops 
print('Number of Nodes with self loops is: ' + str(len(list(nx.nodes_with_selfloops(G2)))))

# Number of Reciprocated Edges (Το documentation του networkx μας αναφέρει ότι η συνάρτηση reciprocity υπολογίζει τον λόγο των ακμών που δείχνουν και στις δύο 
# κατευθήνσεις, προς το πλήθος όλων των ακμών του δικτύου. Οπότε για να υπολογίζουμε το πλήθος των reciprocated edges αρκεί να πολλαπλασιάσουμε το αποτέλεσμα
# που παράγει η συνάρτηση reciprocity με το πλήθος όλων των ακμών) 
print('Number of reciprocated edges is: ' + str(reciprocity(G2)*G2.number_of_edges()))

#Number of Sink and Source nodes (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)

# #Max, Min and Average Degree 
degrees_G2 = dict(G2.degree())
print('Max Degree of a Node: ' + str(max(degrees_G2.values()))) 
print('Min Degree of a Node: ' + str(min(degrees_G2.values()))) 
avg_degree_G2 = sum(val for key, val in degrees_G2.items()) / G2.number_of_nodes()
print('Average Degree of Nodes: ' + str(avg_degree_G2))

#Max, Min and Average In-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Max, Min and Average Out-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Diameter of the Network (δεν είναι συνεκτικά συνδεδεμένο, οπότε θα βρω την μεγαλύτερη διάμετρο από τις συνιστώσες του)
components = list(nx.connected_components(G2))
max_diameter = 0
for component in components:
    diameter = nx.diameter(G2.subgraph(component))
    if diameter > max_diameter:
        max_diameter = diameter
print("The diameter of the graph is:", max_diameter)

#Average clustering coefficient and global clustering coefficient 
clustering_coefficients = nx.clustering(G2)
avg_clustering_coefficient = sum(val for key,val in clustering_coefficients.items())/G2.number_of_nodes()
print('Average Clustering Coefficient is: ' + str(avg_clustering_coefficient)) 

global_clustering_coefficient = nx.transitivity(G2)
print('Global Clustering Coefficient is ' + str(global_clustering_coefficient)) 

#Largest Strongly Connected And Weakly Connected Component 
#(Δεν έχει σημασία αφού ο γράφος είναι μη-κατευθυνόμενος, οπότε αρκεί να βρούμε μόνο την μεγαλύτερη συνιστώσα)
cc_G2 = list(nx.connected_components(G2))
largest_cc_G2 = max(cc_G2, key=len)
print('Number of Nodes in the largest connected Component: ' + str(len(largest_cc_G2))) 
num_edges_largest_cc_G2 = G2.subgraph(largest_cc_G2).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_cc_G2))

Basic Analysis for General Realtivity and Quantum Cosmology Network (G2)

Number of Nodes for G2 is: 5242
Number of Edges for G2 is: 14496
Number of Nodes with self loops is: 12
Number of reciprocated edges is: 0.0
Max Degree of a Node: 81
Min Degree of a Node: 1
Average Degree of Nodes: 5.530713468141931


KeyboardInterrupt: ignored

# Third Section (Erdos renyi random graph)

Θα δημιουργήσω ένα δίκτυο με την βοήθεια του αλγορίθμου erdos- renyi με 5200 κόμβους. Μένει να διαλέξω τιμή για την παράμετρο p, δηλαδή την πιθανότητα που θα έχουν δύο κόμβοι να συνδεθούν μεταξύ τους. Μετά από λίγο ψάγιμο, βρήκα ότι ο αριθμός των expected ακμών, E, υπολογίζεται σύμφωνα από την σχέση $E = p \cdot \frac{n\cdot(n-1)}{2}$, όπου p είναι η πιθανότητα δύο κόμβοι να συνδέονται μεταξύ τους, και n ο αριθμός των κόμβων του δικτύου. Οπότε, εφόσον θέλω δίκτυο με 5200 κόμβους, και ο αριθμός των ακμών θέλω να είναι παρόμοιος με το δίκτυο G2 (δηλαδή Ε = 14496), υπολογίζω το p από την προηγούμενη σχέση και βρίσω ότι $p=\frac{2\cdot E}{n\cdot (n-1)} = \frac{2\cdot 14496}{5200\cdot 5199} = 0.00107$

In [ ]:
# Create an Erdos-Rényi graph
G3 = nx.erdos_renyi_graph(5200, 0.00107)

Αφού έχει δημιουργηθέι το δίκτυο, κάνουμε τις ίδιες μετρήσεις όπως το G2 και παίρνουμε: 

In [ ]:
print('Basic Analysis for Random Erdos - Renyi Network (G3)')
print()
#nodes - edges 
print('Number of Nodes for G3 is: ' + str(G3.number_of_nodes())) 
print('Number of Edges for G3 is: ' + str(G3.number_of_edges())) 

#Nodes with Self Loops 
print('Number of Nodes with self loops is: ' + str(len(list(nx.nodes_with_selfloops(G3)))))

# Number of Reciprocated Edges (Το documentation του networkx μας αναφέρει ότι η συνάρτηση reciprocity υπολογίζει τον λόγο των ακμών που δείχνουν και στις δύο 
# κατευθήνσεις, προς το πλήθος όλων των ακμών του δικτύου. Οπότε για να υπολογίζουμε το πλήθος των reciprocated edges αρκεί να πολλαπλασιάσουμε το αποτέλεσμα
# που παράγει η συνάρτηση reciprocity με το πλήθος όλων των ακμών) 
print('Number of reciprocated edges is: ' + str(reciprocity(G3)*G3.number_of_edges()))

#Number of Sink and Source nodes (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)

# #Max, Min and Average Degree 
degrees_G3 = dict(G3.degree())
print('Max Degree of a Node: ' + str(max(degrees_G3.values()))) 
print('Min Degree of a Node: ' + str(min(degrees_G3.values()))) 
avg_degree_G3 = sum(val for key, val in degrees_G3.items()) / G3.number_of_nodes()
print('Average Degree of Nodes: ' + str(avg_degree_G3))

#Max, Min and Average In-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Max, Min and Average Out-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Diameter of the Network (δεν είναι συνεκτικά συνδεδεμένο, οπότε θα βρω την μεγαλύτερη διάμετρο από τις συνιστώσες του)
components = list(nx.connected_components(G3))
max_diameter = 0
for component in components:
    diameter = nx.diameter(G3.subgraph(component))
    if diameter > max_diameter:
        max_diameter = diameter
print("The diameter of the graph is:", max_diameter)

#Average clustering coefficient and global clustering coefficient 
clustering_coefficients = nx.clustering(G3)
avg_clustering_coefficient = sum(val for key,val in clustering_coefficients.items())/G3.number_of_nodes()
print('Average Clustering Coefficient is: ' + str(avg_clustering_coefficient)) 

global_clustering_coefficient = nx.transitivity(G3)
print('Global Clustering Coefficient is ' + str(global_clustering_coefficient)) 

#Largest Strongly Connected And Weakly Connected Component 
#(Δεν έχει σημασία αφού ο γράφος είναι μη-κατευθυνόμενος, οπότε αρκεί να βρούμε μόνο την μεγαλύτερη συνιστώσα)
cc_G3 = list(nx.connected_components(G3))
largest_cc_G3 = max(cc_G3, key=len)
print('Number of Nodes in the largest connected Component: ' + str(len(largest_cc_G3))) 
num_edges_largest_cc_G3 = G3.subgraph(largest_cc_G3).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_cc_G3))

Basic Analysis for Random Erdos - Renyi Network (G3)

Number of Nodes for G3 is: 5200
Number of Edges for G3 is: 14453
Number of Nodes with self loops is: 0
Number of reciprocated edges is: 0.0
Max Degree of a Node: 15
Min Degree of a Node: 0
Average Degree of Nodes: 5.5588461538461535


KeyboardInterrupt: ignored

Παρατηρούμε ότι τα χαρακτηριστικά του γράφου που δημιουργήθηκε είναι παρόμοια με τα χαρακτηριστικά του δεύτερου δικτύου, με εξαίρεση τον μέγιστο βαθμό των κόμβων, ο οποίος στο δευ΄τερο δίκτυο ήταν 81 ενώ τώρα είναι μόνο 15, και το clustering coefficient (average και global), το οποίο έπεσε κατά πολύ. Αν μπορούσα να το ερμηνεύσω θα έλεγα ότι είναι λογικό να συμβαίνει κάτι τέτοιο, αφού αν ένας κόμβος έχει πιθανότητα 0.00107 να συνδεθεί με κάποιον γείτονά του, η πιθανότητα να συνδεθέι ένας ακόμα γείτονας με τον ίδιο κόμβο πέφτει στο $0.00107\cdot 0.00107$, κάνοντας έτσι την τιμή του clustering coefficient για κάθε κόμβο να γίνει πολύ μικρή.

# Fourth Section (Watts και Strogatz random Graph)

Το να δημιουργηθεί δίκτυο χρησιμοποιώντας τον αλγόριθμο Watts and Strogatz είναι πιο εύκολο όσον αφορά τις επιλογές των παραμέτρων ώστε να μοιάζει το δεύτερο δίκτυο. Πιο συγκεκριμένα έχουμε να επιλέξουμε 3 τιμές, μία για τους κόμβους που ξέρουμε ήδη ότι είναι 5200, μία για το k, το οποίο είναι ο αριθμός των γειτόνων που συνδέεται ένας κομβός μαζί τους, και μία για το p, το οποίο είναι η πιθανότητα μία ακμή από αυτές που συνδέεουν τους k πλησιέστερους γείτονες με έναν κόμβο να διαγραφεί και να δημιουργηθεί μία νέα. Οπότε, εφόσον ο μέσος βαθμός των κόμβων στο δεύτερο δίκτυο ήταν 5.5307, θέτω το k = 6, δηλαδή κάθε κόμβος θα συνδέεται με τους 6 κοντινότερους γείτονές του. Και επειδή θέλω να διατηρηθεί το clustering coefficient σχετικά ψηλά, θέλω οι γείτονες ενός κόμβου να συνδέονται και μεταξύ τους, οπότε διαλέγω μία μικ΄ρη τιμή για το p, p = 0.05.  

In [ ]:
G4 = nx.watts_strogatz_graph(5200, 6, 0.05)

In [ ]:
print('Basic Analysis for Random Watts and Strogatz Graph (G4)')
print()
#nodes - edges 
print('Number of Nodes for G4 is: ' + str(G4.number_of_nodes())) 
print('Number of Edges for G4 is: ' + str(G4.number_of_edges())) 

#Nodes with Self Loops 
print('Number of Nodes with self loops is: ' + str(len(list(nx.nodes_with_selfloops(G4)))))

# Number of Reciprocated Edges (Το documentation του networkx μας αναφέρει ότι η συνάρτηση reciprocity υπολογίζει τον λόγο των ακμών που δείχνουν και στις δύο 
# κατευθήνσεις, προς το πλήθος όλων των ακμών του δικτύου. Οπότε για να υπολογίζουμε το πλήθος των reciprocated edges αρκεί να πολλαπλασιάσουμε το αποτέλεσμα
# που παράγει η συνάρτηση reciprocity με το πλήθος όλων των ακμών) 
print('Number of reciprocated edges is: ' + str(reciprocity(G4)*G4.number_of_edges()))

#Number of Sink and Source nodes (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)

# #Max, Min and Average Degree 
degrees_G4 = dict(G4.degree())
print('Max Degree of a Node: ' + str(max(degrees_G4.values()))) 
print('Min Degree of a Node: ' + str(min(degrees_G4.values()))) 
avg_degree_G4 = sum(val for key, val in degrees_G4.items()) / G4.number_of_nodes()
print('Average Degree of Nodes: ' + str(avg_degree_G4))

#Max, Min and Average In-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Max, Min and Average Out-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Diameter of the Network (δεν είναι συνεκτικά συνδεδεμένο, οπότε θα βρω την μεγαλύτερη διάμετρο από τις συνιστώσες του)
components = list(nx.connected_components(G4))
max_diameter = 0
for component in components:
    diameter = nx.diameter(G4.subgraph(component))
    if diameter > max_diameter:
        max_diameter = diameter
print("The diameter of the graph is:", max_diameter)

#Average clustering coefficient and global clustering coefficient 
clustering_coefficients = nx.clustering(G4)
avg_clustering_coefficient = sum(val for key,val in clustering_coefficients.items())/G4.number_of_nodes()
print('Average Clustering Coefficient is: ' + str(avg_clustering_coefficient)) 

global_clustering_coefficient = nx.transitivity(G4)
print('Global Clustering Coefficient is ' + str(global_clustering_coefficient)) 

#Largest Strongly Connected And Weakly Connected Component 
#(Δεν έχει σημασία αφού ο γράφος είναι μη-κατευθυνόμενος, οπότε αρκεί να βρούμε μόνο την μεγαλύτερη συνιστώσα)
cc_G4 = list(nx.connected_components(G4))
largest_cc_G4 = max(cc_G4, key=len)
print('Number of Nodes in the largest connected Component: ' + str(len(largest_cc_G4))) 
num_edges_largest_cc_G4 = G4.subgraph(largest_cc_G4).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_cc_G4))

Basic Analysis for Random Watts and Strogatz Graph (G4)

Number of Nodes for G4 is: 5200
Number of Edges for G4 is: 15600
Number of Nodes with self loops is: 0
Number of reciprocated edges is: 0.0
Max Degree of a Node: 8
Min Degree of a Node: 3
Average Degree of Nodes: 6.0


KeyboardInterrupt: ignored

Βλέπουμε ότι όντως τα χαρακτηριστικά του δεύτερου και του τελευταίου δικτύου είναι παρόμοια.

# Fifth Section (Barabasi - Albert Graph - Undirected) 

Εδώ πρέπει να βρούμε την κατάλληλη τιμή για το m, όπου m είναι ο αριθμος των κόμβων που συνδέεται ο κόμβος που εισάγεται κάθε φορά στο δίκτυο. Εδώ στην αρχή είχα διαλέξει για το m την τιμή 6, καθώς είχα σκεφτεί ότι εφόσον έχουμε μέσο βαθμό κόμβων το 6, κάθε κόμβος που εισάγεται θα πρέπει να συνδεθεί με 6 κόμβους οι οποίοι προυπήρχαν στο δίκτυο. Με αυτόν τον τρόπο όμως, οι τελικές ακμές του δικτύου έβγαιναν περίπου 32000, κάτι το οποίο μου φάνηκε παράξενο, όπως και το γεγονός ότι ο μέσος βαθμών των κόμβων έβγαινε περίπου 11. Μετά απο σκέψη όμως συνειδητοποίησα ότι όταν ξεκινάει να δημιουργείται το δίκτυο, αν προσθέτω κάθε φορά 6 ακμές με κάθε εισαγωγή κόμβου, το δίκτυο γίνεται υπερβολικά πολύ πυκνό, μεγαλόνωντας κατά πολύ έτσι τον μέσο βαθμό των κόμβων. Επίσης σε αυτό το μοντέλο συνειδητοποίησα ότι είναι λογικό ο μέγιστος βαθμός των κόμβων να είναι μεγάλος, λόγω και πάλι του ξεκινήματος δημιουργίας του δικτύου, αφού οι νέοι κόμβοι που εισέρχονται δεν έχουν πολλές επιλογές για να συνδεθούν με άλλους κόμβους. Για να διορθωθεί το πρόβλημα των πολλών ακμών επίσης, χρησιμοποίησα m = 3, για να πλησιάσουμε τις 15000 ακμές.

In [ ]:
G5 = nx.barabasi_albert_graph(5200, 3)


In [ ]:
print('Basic Analysis for Random Barabasi - Albert Graph (G5)')
print()
#nodes - edges 
print('Number of Nodes for G5 is: ' + str(G5.number_of_nodes())) 
print('Number of Edges for G5 is: ' + str(G5.number_of_edges())) 

#Nodes with Self Loops 
print('Number of Nodes with self loops is: ' + str(len(list(nx.nodes_with_selfloops(G5)))))

# Number of Reciprocated Edges (Το documentation του networkx μας αναφέρει ότι η συνάρτηση reciprocity υπολογίζει τον λόγο των ακμών που δείχνουν και στις δύο 
# κατευθήνσεις, προς το πλήθος όλων των ακμών του δικτύου. Οπότε για να υπολογίζουμε το πλήθος των reciprocated edges αρκεί να πολλαπλασιάσουμε το αποτέλεσμα
# που παράγει η συνάρτηση reciprocity με το πλήθος όλων των ακμών) 
print('Number of reciprocated edges is: ' + str(reciprocity(G5)*G5.number_of_edges()))

#Number of Sink and Source nodes (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)

# #Max, Min and Average Degree 
degrees_G5 = dict(G5.degree())
print('Max Degree of a Node: ' + str(max(degrees_G5.values()))) 
print('Min Degree of a Node: ' + str(min(degrees_G5.values()))) 
avg_degree_G5 = sum(val for key, val in degrees_G5.items()) / G5.number_of_nodes()
print('Average Degree of Nodes: ' + str(avg_degree_G5))

#Max, Min and Average In-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Max, Min and Average Out-going Degree (Δεν βγάζει νόημα αφού έχουμε μη-κατευθυνόμενο γράφο)


#Diameter of the Network (δεν είναι συνεκτικά συνδεδεμένο, οπότε θα βρω την μεγαλύτερη διάμετρο από τις συνιστώσες του)
components = list(nx.connected_components(G5))
max_diameter = 0
for component in components:
    diameter = nx.diameter(G5.subgraph(component))
    if diameter > max_diameter:
        max_diameter = diameter
print("The diameter of the graph is:", max_diameter)

#Average clustering coefficient and global clustering coefficient 
clustering_coefficients = nx.clustering(G5)
avg_clustering_coefficient = sum(val for key,val in clustering_coefficients.items())/G5.number_of_nodes()
print('Average Clustering Coefficient is: ' + str(avg_clustering_coefficient)) 

global_clustering_coefficient = nx.transitivity(G5)
print('Global Clustering Coefficient is ' + str(global_clustering_coefficient)) 

#Largest Strongly Connected And Weakly Connected Component 
#(Δεν έχει σημασία αφού ο γράφος είναι μη-κατευθυνόμενος, οπότε αρκεί να βρούμε μόνο την μεγαλύτερη συνιστώσα)
cc_G5 = list(nx.connected_components(G5))
largest_cc_G5 = max(cc_G5, key=len)
print('Number of Nodes in the largest connected Component: ' + str(len(largest_cc_G5))) 
num_edges_largest_cc_G5 = G5.subgraph(largest_cc_G5).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_cc_G5))

Basic Analysis for Random Barabasi - Albert Graph (G5)

Number of Nodes for G5 is: 5200
Number of Edges for G5 is: 15591
Number of Nodes with self loops is: 0
Number of reciprocated edges is: 0.0
Max Degree of a Node: 220
Min Degree of a Node: 3
Average Degree of Nodes: 5.996538461538462


KeyboardInterrupt: ignored

Παρατηρούμε για ακόμα μία φορά ότι το clustering coefficient είναι πιο μικρό από ότι θα έπρεπε, και αυτό είναι λογικό να συμβαίνει διότι οι γείτονες κάθε νέου κόμβου που προστίθεται δεν είναι πολύ εύκολο να συνδέονται μεταξύ τους. 

# Sixth Section (Baraasi - Albert Graph - Directed)  

Εδώ πέρα δεν κατάλαβα πως ένας κατευθυνόμενος γράφος μπορεί να έχει τα ίδια χαρακτηριστικά με κάποιον μη- κατευθυνόμενο, οπότε δημιοούργησα με τον αλγόριθμο Barabasi - Albert ενα δίκτυο με όλες τις ακμές του reciprocated, ώστε να δημιουργηθεί ένα πλήθος ακμών που να μπορεί να συγκριθεί με το δεύτερο δίκτυο. 

In [ ]:
G6 = nx.DiGraph(nx.barabasi_albert_graph(5200, 2))

In [ ]:
print('Basic Analysis for Barabasi - Albert Graph (G6)')
print()
#nodes - edges 
print('Number of Nodes for G6 is: ' + str(G6.number_of_nodes())) 
print('Number of Edges for G6 is: ' + str(G6.number_of_edges())) 

#Nodes with Self Loops 
print('Number of Nodes with self loops is: ' + str(len(list(nx.nodes_with_selfloops(G6)))))

# Number of Reciprocated Edges (Το documentation του networkx μας αναφέρει ότι η συνάρτηση reciprocity υπολογίζει τον λόγο των ακμών που δείχνουν και στις δύο 
# κατευθήνσεις, προς το πλήθος όλων των ακμών του δικτύου. Οπότε για να υπολογίζουμε το πλήθος των reciprocated edges αρκεί να πολλαπλασιάσουμε το αποτέλεσμα
# που παράγει η συνάρτηση reciprocity με το πλήθος όλων των ακμών) 
print('Number of reciprocated edges is: ' + str(reciprocity(G6)*G6.number_of_edges()))

# #Number of Sink and Source nodes 
print('Number of Sink Nodes in G6: ' + str(len([node for node in G6.nodes if G6.out_degree(node) == 0])))
print('Number of Source Nodes in G6: ' + str(len([node for node in G6.nodes if G6.in_degree(node) == 0])))

#Max, Min and Average In-Going Degrees
ingoing_degrees_G6 = dict(G6.in_degree())
avg_ingoing_degree_G6 = sum(val for key, val in ingoing_degrees_G6.items()) / G6.number_of_nodes()
print('Max in-going Degree of a Node: ' + str(max(ingoing_degrees_G6.values()))) 
print('Min in-going Degree of a Node: ' + str(min(ingoing_degrees_G6.values())))
print('Average in-going degree of Nodes: ' + str(avg_ingoing_degree_G6)) 

# #Max, Min and Average Out-going Degree
outgoing_degrees_G6 = dict(G1.out_degree())
avg_outgoing_degree_G6 = sum(val for key, val in outgoing_degrees_G6.items()) / G6.number_of_nodes()
print('Max Out-going Degree of a Node: ' + str(max(outgoing_degrees_G6.values()))) 
print('Min Out-going Degree of a Node: ' + str(min(outgoing_degrees_G6.values())))
print('Average Out-going degree of Nodes: ' + str(avg_outgoing_degree_G6)) 

#Diameter of the Network (δεν είναι συνεκτικά συνδεδεμένο, οπότε θα βρω την μεγαλύτερη διάμετρο από τις συνιστώσες του. Επίσης δεν λαμβάνω υπόψη την φορά των συνιστωσών του.
components = list(nx.weakly_connected_components(G6))
max_diameter = 0
for component in components:
    diameter = nx.diameter(G6.subgraph(component).to_undirected())
    if diameter > max_diameter:
        max_diameter = diameter
print("The diameter is:", max_diameter)

#Average clustering coefficient and global clustering coefficient 
clustering_coefficients = nx.clustering(G6)
avg_clustering_coefficient = sum(val for key,val in clustering_coefficients.items())/G6.number_of_nodes()
print('Average Clustering Coefficient is: ' + str(avg_clustering_coefficient*2)) 
#Για να βγει το σωστό αποτέλεσμα το πολλαπλασιάζω *2 

global_clustering_coefficient = nx.transitivity(G6)
print('Global Clustering Coefficient is ' + str(global_clustering_coefficient*2)) 

#strongly connected component
scc_G6 = list(nx.strongly_connected_components(G6))
largest_scc_G6 = max(scc_G6, key=len)
print('Number of Nodes in the largest Strongly connected Component: ' + str(len(largest_scc_G6))) 
num_edges_largest_scc_G6 = G6.subgraph(largest_scc_G6).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_scc_G6))

#Weakly connected component  
wcc_G6 = list(nx.weakly_connected_components(G6))

largest_wcc_G6 = max(wcc_G6, key=len)
print('Number of Nodes in the largest Weakly connected Component: ' + str(len(largest_wcc_G6))) 
num_edges_largest_wcc_G6 = G6.subgraph(largest_wcc_G6).number_of_edges()
print('Number of Nodes in the largest Strongly connected Component: ' + str(num_edges_largest_wcc_G6))


# Συγκριτικά Αποτελέσματα όλων των δικτύων αποτυπωμένα σε πίνακα 


|  | Gnutella (G1) |GrQc (G2)| Erdős-Renyi (G3) |Watts and Strogatz (G4)|Barabasi - Albert (Undirected) (G5) |Barabasi - Albert (Directed) (G6)|
|-------|:-----|:----- |:-----|:-----|:-----|:------|
| Number Of Nodes| 8846 | 5242| 5200|5200|5200|5200
| Number Of Nodes with Self Loops|0|12| 0|0|0|0
|Number of Edges|31839|14496| 14567| 15600|15591|20792
|Number of Recirpocated Edges| 0|0| 0| 0 |0|20792
|Number of Sink Nodes|4996|-| - |-|-|0
|Number of Source Nodes|118|-| - |-|-|0
|Max Degree|-| 81| 15|10|264|-
|Min Degree|-| 1| 0 |4|3|-
|Average Degree|-| 5.5307| 5.6026|6|5.9965|-
|Max in-going Degree|79|-| -|-|-|133
|Min in-going Degree|0|-| -|-|-|2
|Average in-going Degree|3.599|-| -| -|-|3.9984
|Max out-going Degree|65|-| - |-|-|65
|Min out-going Degree|0|-|-|-|-|0
|Average out-going Degree|3.599|-|-|-|-|6.122
|Diameter|9|17|10 |20|7|8
|Average Clustering Coefficient| 0.007201|0.5296| 0.00100| 0.5218|0.01060|0.01274
|Global Clustering Coefficient|0.007493|0.62984|0.00102 |0.5134|0.005266|0.00464
|Number of Nodes in Largest Strongly Connected Component| 3234| 4158|5179 |5200|5200|5200
|Number of Edges in Largest Strongly Connected Component|13453|13428|14567 |15600|15591|20792
|Number of Nodes in Largest Weakly Connected Component|8842|4158|5179 |5200|5200|5200
|Number of Edges in Largest Weakly Connected Component|31837|13428| 14567|15600|15591|20792

# Edge Distribution for every Graph 

### Edge distribution using Linear Scale 

In [ ]:
#Θα πρέπει να δημιουργηθεί ένα λεξικό το οποίο θα μετράει για κάθε δίκτυο την πιθανότητα των ακμών. 
#Για τα κατευθυνόμενα δίκτυα θα δημιουργήσω δύο γραφήματα, ένα για τις εξερχόμενες και ένα για τις εισερχόμενες ακμές

#G1
#In-going Degrees
degree_in_count_G1 = {}
for d in ingoing_degrees_G1.values():
    if d not in degree_in_count_G1:
        degree_in_count_G1[d] = 1
    else:
        degree_in_count_G1[d] += 1
total_nodes = G1.number_of_nodes()
degree_in_prob_G1 = {k: v / total_nodes for k, v in degree_in_count_G1.items()}
# Plot the degree distribution
plt.scatter(degree_in_prob_G1.keys(), degree_in_prob_G1.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Ingoing-Degree Distribution of G1 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

#Outgoing Degrees
degree_out_count_G1 = {}
for d in outgoing_degrees_G1.values():
    if d not in degree_out_count_G1:
        degree_out_count_G1[d] = 1
    else:
        degree_out_count_G1[d] += 1
total_nodes = G1.number_of_nodes()
degree_out_prob_G1 = {k: v / total_nodes for k, v in degree_out_count_G1.items()}
# Plot the degree distribution
plt.scatter(degree_out_prob_G1.keys(), degree_out_prob_G1.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Outgoing-Degree Distribution of G1 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

#G2
degree_count_G2 = {}
for d in degrees_G2.values():
    if d not in degree_count_G2:
        degree_count_G2[d] = 1
    else:
        degree_count_G2[d] += 1
total_nodes = G2.number_of_nodes()
degree_prob_G2 = {k: v / total_nodes for k, v in degree_count_G2.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G2.keys(), degree_prob_G2.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G2 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

#G3
degree_count_G3 = {}
for d in degrees_G3.values():
    if d not in degree_count_G3:
        degree_count_G3[d] = 1
    else:
        degree_count_G3[d] += 1
total_nodes = G3.number_of_nodes()
degree_prob_G3 = {k: v / total_nodes for k, v in degree_count_G3.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G3.keys(), degree_prob_G3.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G3 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

#G4
degree_count_G4 = {}
for d in degrees_G4.values():
    if d not in degree_count_G4:
        degree_count_G4[d] = 1
    else:
        degree_count_G4[d] += 1
total_nodes = G4.number_of_nodes()
degree_prob_G4 = {k: v / total_nodes for k, v in degree_count_G4.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G4.keys(), degree_prob_G4.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G4 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

#G5
degree_count_G5 = {}
for d in degrees_G5.values():
    if d not in degree_count_G5:
        degree_count_G5[d] = 1
    else:
        degree_count_G5[d] += 1
total_nodes = G5.number_of_nodes()
degree_prob_G5 = {k: v / total_nodes for k, v in degree_count_G5.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G5.keys(), degree_prob_G5.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G5 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

#G6
#In-going Degrees
degree_in_count_G6 = {}
for d in ingoing_degrees_G6.values():
    if d not in degree_in_count_G6:
        degree_in_count_G6[d] = 1
    else:
        degree_in_count_G6[d] += 1
total_nodes = G6.number_of_nodes()
degree_in_prob_G6 = {k: v / total_nodes for k, v in degree_in_count_G6.items()}
# Plot the degree distribution
plt.scatter(degree_in_prob_G6.keys(), degree_in_prob_G6.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Ingoing-Degree Distribution of G6 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()
#Outgoing Degrees
degree_out_count_G6 = {}
for d in outgoing_degrees_G6.values():
    if d not in degree_out_count_G6:
        degree_out_count_G6[d] = 1
    else:
        degree_out_count_G6[d] += 1
total_nodes = G6.number_of_nodes()
degree_out_prob_G6 = {k: v / total_nodes for k, v in degree_out_count_G6.items()}
# Plot the degree distribution
plt.scatter(degree_out_prob_G6.keys(), degree_out_prob_G6.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Outgoing-Degree Distribution of G6 Graph')
plt.xscale('linear')
plt.yscale('linear')
plt.show()

### Edge distribution using Log-Log Scale

In [ ]:
#Με παρόμοια κώδικα όπως πριν θα δημιουργήσω και τα γραφήματα σε λογαριθμική κλίμακα, απλά θα θέσω το x.scale και το y.scale να είναι log
#G1
#In-going Degrees
degree_in_count_G1 = {}
for d in ingoing_degrees_G1.values():
    if d not in degree_in_count_G1:
        degree_in_count_G1[d] = 1
    else:
        degree_in_count_G1[d] += 1
total_nodes = G1.number_of_nodes()
degree_in_prob_G1 = {k: v / total_nodes for k, v in degree_in_count_G1.items()}
# Plot the degree distribution
plt.scatter(degree_in_prob_G1.keys(), degree_in_prob_G1.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Ingoing-Degree Distribution of G1 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

#Outgoing Degrees
degree_out_count_G1 = {}
for d in outgoing_degrees_G1.values():
    if d not in degree_out_count_G1:
        degree_out_count_G1[d] = 1
    else:
        degree_out_count_G1[d] += 1
total_nodes = G1.number_of_nodes()
degree_out_prob_G1 = {k: v / total_nodes for k, v in degree_out_count_G1.items()}
# Plot the degree distribution
plt.scatter(degree_out_prob_G1.keys(), degree_out_prob_G1.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Outgoing-Degree Distribution of G1 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

#G2
degree_count_G2 = {}
for d in degrees_G2.values():
    if d not in degree_count_G2:
        degree_count_G2[d] = 1
    else:
        degree_count_G2[d] += 1
total_nodes = G2.number_of_nodes()
degree_prob_G2 = {k: v / total_nodes for k, v in degree_count_G2.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G2.keys(), degree_prob_G2.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G2 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

#G3
degree_count_G3 = {}
for d in degrees_G3.values():
    if d not in degree_count_G3:
        degree_count_G3[d] = 1
    else:
        degree_count_G3[d] += 1
total_nodes = G3.number_of_nodes()
degree_prob_G3 = {k: v / total_nodes for k, v in degree_count_G3.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G3.keys(), degree_prob_G3.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G3 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

#G4
degree_count_G4 = {}
for d in degrees_G4.values():
    if d not in degree_count_G4:
        degree_count_G4[d] = 1
    else:
        degree_count_G4[d] += 1
total_nodes = G4.number_of_nodes()
degree_prob_G4 = {k: v / total_nodes for k, v in degree_count_G4.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G4.keys(), degree_prob_G4.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G4 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

#G5
degree_count_G5 = {}
for d in degrees_G5.values():
    if d not in degree_count_G5:
        degree_count_G5[d] = 1
    else:
        degree_count_G5[d] += 1
total_nodes = G5.number_of_nodes()
degree_prob_G5 = {k: v / total_nodes for k, v in degree_count_G5.items()}
# Plot the degree distribution
plt.scatter(degree_prob_G5.keys(), degree_prob_G5.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Degree Distribution of G5 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

#G6
#In-going Degrees
degree_in_count_G6 = {}
for d in ingoing_degrees_G6.values():
    if d not in degree_in_count_G6:
        degree_in_count_G6[d] = 1
    else:
        degree_in_count_G6[d] += 1
total_nodes = G6.number_of_nodes()
degree_in_prob_G6 = {k: v / total_nodes for k, v in degree_in_count_G6.items()}
# Plot the degree distribution
plt.scatter(degree_in_prob_G6.keys(), degree_in_prob_G6.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Ingoing-Degree Distribution of G6 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()
#Outgoing Degrees
degree_out_count_G6 = {}
for d in outgoing_degrees_G6.values():
    if d not in degree_out_count_G6:
        degree_out_count_G6[d] = 1
    else:
        degree_out_count_G6[d] += 1
total_nodes = G6.number_of_nodes()
degree_out_prob_G6 = {k: v / total_nodes for k, v in degree_out_count_G6.items()}
# Plot the degree distribution
plt.scatter(degree_out_prob_G6.keys(), degree_out_prob_G6.values())
plt.xlabel('Degree')
plt.ylabel('Probability')
plt.title('Outgoing-Degree Distribution of G6 Graph using log-log scale')
plt.xscale('log')
plt.yscale('log')
plt.show()

### Edge distribution using Log - Bining Scale

In [ ]:
#Πρώτα από όλα βρίσκω όπως το αρχείο που ανέβηκε στο open e-class, τα kmin και kmax κάθε δικτύου, και σχηματίζω σε λίστα τους βαθμούς των δικτύων. Στην περίπτωση των κατευθυνόμενων γράφων θα δημιουργήσω
# δύο γραφήματα για κάθε περίπτωση, ένα για την κατανομή των εξερχόμενων ακμών, και ένα για την κατανομή των εισερχόμενων ακμών

#Gnutella Network
in_degrees_G1 = [d for n, d in G1.in_degree()]
k_in_min_G1 = min(in_degrees_G1)
k_in_max_G1 = max(in_degrees_G1)

out_degrees_G1 = [d for n, d in G1.out_degree()]
k_out_min_G1 = min(out_degrees_G1)
k_out_max_G1 = max(out_degrees_G1)

# G2
degrees_G2 = [d for n, d in G2.degree()]
k_min_G2 = min(degrees_G2)
k_max_G2 = max(degrees_G2)

#G3 
degrees_G3 = [d for n, d in G3.degree()]
k_min_G3 = min(degrees_G3)
k_max_G3 = max(degrees_G3)

#G4
degrees_G4 = [d for n, d in G4.degree()]
k_min_G4 = min(degrees_G4)
k_max_G4 = max(degrees_G4)

#G5
degrees_G5 = [d for n, d in G5.degree()]
k_min_G5 = min(degrees_G5)
k_max_G5 = max(degrees_G5)

#G6
in_degrees_G6 = [d for n, d in G6.in_degree()]
k_in_min_G6 = min(in_degrees_G6)
k_in_max_G6 = max(in_degrees_G6)

out_degrees_G6 = [d for n, d in G6.out_degree()]
k_out_min_G6 = min(out_degrees_G6)
k_out_max_G6 = max(out_degrees_G6)


In [ ]:
#G1
#In-going
bins = np.logspace(np.log10(k_in_min_G1+1), np.log10(k_in_max_G1+1), 10)
hist, _ = np.histogram(in_degrees_G1, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for in-going degrees of G1')
plt.show()
#Out-going
bins = np.logspace(np.log10(k_out_min_G1+1), np.log10(k_out_max_G1+1), 10)
hist, _ = np.histogram(out_degrees_G1, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for out-going degrees of G1')
plt.show()

#G2
bins = np.logspace(np.log10(k_min_G2), np.log10(k_max_G2+1), 10)
hist, _ = np.histogram(degrees_G2, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for G2')
plt.show()

#G3
bins = np.logspace(np.log10(k_min_G3+1), np.log10(k_max_G3+1), 10)
hist, _ = np.histogram(degrees_G3, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for G3')
plt.show()

#G4
bins = np.logspace(np.log10(k_min_G4), np.log10(k_max_G4+1), 10)
hist, _ = np.histogram(degrees_G4, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for G4')
plt.show()

#G5
bins = np.logspace(np.log10(k_min_G5), np.log10(k_max_G5+1), 10)
hist, _ = np.histogram(degrees_G5, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for G5')
plt.show()

#G6
#In-going
bins = np.logspace(np.log10(k_in_min_G6+1), np.log10(k_in_max_G6+1), 10)
hist, _ = np.histogram(in_degrees_G6, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for in-going degrees of G6')
plt.show()
#Out-going
bins = np.logspace(np.log10(k_out_min_G6+1), np.log10(k_out_max_G6+1), 10)
hist, _ = np.histogram(out_degrees_G6, bins=bins)
bin_centers = (bins[1:] + bins[:-1])/2
pdf = hist/sum(hist)
# Plot the degree distribution using logarithmic bins
fig, ax = plt.subplots()
ax.loglog(bin_centers, pdf, '.', markersize=10)
ax.set_xlabel('Degree')
ax.set_ylabel('Probability ')
ax.set_title('Degree Distribution (Logarithmic Binning) for out-going degrees of G6')
plt.show()
